// ng-lightning/forms/groups/group-alt.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, Input, ChangeDetectionStrategy, HostBinding, ContentChild } from '@angular/core';
import { NglFormGroup } from './group';
import { NglFormLabelTemplate } from '../form-label';
var NglFormGroupAlternate = (function (_super) {
    __extends(NglFormGroupAlternate, _super);
    function NglFormGroupAlternate() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    // AoT workaround
    NglFormGroupAlternate.prototype.ngAfterContentInit = function () {
        _super.prototype.ngAfterContentInit.call(this);
    };
    NglFormGroupAlternate.decorators = [
        { type: Component, args: [{
                    selector: 'fieldset[ngl-form-group-alt]',
                    template: "\n    <legend class=\"slds-form-element__label\"><abbr class=\"slds-required\" *ngIf=\"required\" title=\"required\">*</abbr><span [nglInternalOutlet]=\"_label\"></span></legend><div class=\"slds-form-element__control\"><div [ngClass]=\"'slds-' + type + '--button-group'\"><ng-content></ng-content></div></div><div class=\"slds-form-element__help\" *ngIf=\"error\">{{error}}</div>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        '[class.slds-form-element]': 'true',
                    },
                },] },
    ];
    /** @nocollapse */
    NglFormGroupAlternate.ctorParameters = function () { return []; };
    NglFormGroupAlternate.propDecorators = {
        'labelStr': [{ type: Input, args: ['label',] },],
        'labelTpl': [{ type: ContentChild, args: [NglFormLabelTemplate,] },],
        'error': [{ type: HostBinding, args: ['class.slds-has-error',] }, { type: Input },],
        'required': [{ type: Input },],
        'type': [{ type: Input },],
    };
    return NglFormGroupAlternate;
}(NglFormGroup));
export { NglFormGroupAlternate };
;
