// ng-lightning/forms/elements/checkbox-toggle.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, Input, ChangeDetectionStrategy, ContentChild, ChangeDetectorRef } from '@angular/core';
import { NglFormElement } from './element';
import { NglFormCheckbox } from './input';
import { NglFormLabelTemplate } from '../form-label';
var NglFormElementCheckboxToggle = (function (_super) {
    __extends(NglFormElementCheckboxToggle, _super);
    function NglFormElementCheckboxToggle(detector) {
        var _this = _super.call(this, detector) || this;
        _this.enabledText = 'Enabled';
        _this.disabledText = 'Disabled';
        return _this;
    }
    // AoT workaround
    NglFormElementCheckboxToggle.prototype.ngAfterContentInit = function () {
        _super.prototype.ngAfterContentInit.call(this);
    };
    NglFormElementCheckboxToggle.decorators = [
        { type: Component, args: [{
                    selector: 'ngl-form-checkbox-toggle',
                    template: "\n    <label class=\"slds-checkbox--toggle slds-grid\"><abbr class=\"slds-required\" *ngIf=\"required\" title=\"required\">*</abbr><span class=\"slds-form-element__label slds-m-bottom--none\" [nglInternalOutlet]=\"_label\"></span><ng-content></ng-content><span class=\"slds-checkbox--faux_container\" [attr.for]=\"uid\" aria-live=\"assertive\"><span class=\"slds-checkbox--faux\"></span><span class=\"slds-checkbox--on\">{{enabledText}}</span><span class=\"slds-checkbox--off\">{{disabledText}}</span></span></label><div class=\"slds-form-element__help\" *ngIf=\"error\" [id]=\"'error_' + uid\">{{error}}</div>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        '[class.slds-form-element]': 'true',
                        '[class.slds-has-error]': '!!error',
                    },
                    providers: [{ provide: NglFormElement, useExisting: NglFormElementCheckboxToggle }],
                    styles: [":host { display: block; }"],
                },] },
    ];
    /** @nocollapse */
    NglFormElementCheckboxToggle.ctorParameters = function () { return [
        { type: ChangeDetectorRef, },
    ]; };
    NglFormElementCheckboxToggle.propDecorators = {
        'contentEl': [{ type: ContentChild, args: [NglFormCheckbox,] },],
        'labelStr': [{ type: Input, args: ['label',] },],
        'labelTpl': [{ type: ContentChild, args: [NglFormLabelTemplate,] },],
        'error': [{ type: Input },],
        'enabledText': [{ type: Input },],
        'disabledText': [{ type: Input },],
    };
    return NglFormElementCheckboxToggle;
}(NglFormElement));
export { NglFormElementCheckboxToggle };
;
