// ng-lightning/forms/elements/checkbox.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, Input, ChangeDetectionStrategy, ContentChild, ChangeDetectorRef } from '@angular/core';
import { NglFormElement } from './element';
import { NglFormCheckbox } from './input';
import { NglFormLabelTemplate } from '../form-label';
var NglFormElementCheckbox = (function (_super) {
    __extends(NglFormElementCheckbox, _super);
    function NglFormElementCheckbox(detector) {
        return _super.call(this, detector) || this;
    }
    // AoT workaround
    NglFormElementCheckbox.prototype.ngAfterContentInit = function () {
        _super.prototype.ngAfterContentInit.call(this);
    };
    NglFormElementCheckbox.decorators = [
        { type: Component, args: [{
                    selector: 'ngl-form-checkbox',
                    template: "\n    <div class=\"slds-form-element__control\"><span class=\"slds-checkbox\"><abbr class=\"slds-required\" *ngIf=\"required\" title=\"required\">*</abbr><ng-content></ng-content><label class=\"slds-checkbox__label\" [attr.for]=\"uid\"><span class=\"slds-checkbox--faux\"></span><span class=\"slds-form-element__label\" [nglInternalOutlet]=\"_label\"></span></label></span></div><div class=\"slds-form-element__help\" *ngIf=\"error\" [id]=\"'error_' + uid\">{{error}}</div>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        '[class.slds-form-element]': 'true',
                        '[class.slds-has-error]': '!!error',
                    },
                    providers: [{ provide: NglFormElement, useExisting: NglFormElementCheckbox }],
                    styles: [":host { display: block; }"],
                },] },
    ];
    /** @nocollapse */
    NglFormElementCheckbox.ctorParameters = function () { return [
        { type: ChangeDetectorRef, },
    ]; };
    NglFormElementCheckbox.propDecorators = {
        'contentEl': [{ type: ContentChild, args: [NglFormCheckbox,] },],
        'labelStr': [{ type: Input, args: ['label',] },],
        'labelTpl': [{ type: ContentChild, args: [NglFormLabelTemplate,] },],
        'error': [{ type: Input },],
    };
    return NglFormElementCheckbox;
}(NglFormElement));
export { NglFormElementCheckbox };
;
